src/ActionsLinux.cpp

```diff
@@ -0,0 +1,210 @@
+#include <fcntl.h>
+
+#include "Actions.hpp"
+#include "MemAllocHeap.hpp"
+#include "StatCache.hpp"
+#include "BinaryData.hpp"
+#include "MemAllocLinear.hpp"
+
+#if defined(TUNDRA_LINUX)
+
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <linux/fs.h>
+#ifdef FICLONE
+#include <sys/ioctl.h>
+#endif
+
+ExecResult CopyFiles(const FrozenFileAndHash* src_files, const FrozenFileAndHash* target_files, size_t files_count, StatCache* stat_cache, MemAllocHeap* heap)
+{
+    ExecResult result;
+    memset(&result, 0, sizeof(result));
+    char tmpBuffer[1024];
+
+    int in_file = -1, out_file = -1;
+    int temporary_pipe[2] = { -1, -1 };
+    pipe(temporary_pipe);
+    const size_t splice_chunk_size = 4096;
+
+    MemAllocLinear scratch;
+    LinearAllocInit(&scratch, heap, 4096, "CopyFiles scratch memory");
+
+    for(size_t i = 0; i < files_count; ++i)
+    {
+        LinearAllocReset(&scratch);
+
+        const char* src_file = src_files[i].m_Filename;
+        const char* target_file = target_files[i].m_Filename;
+
+        // The StatCache is not used for the file info because we need full stat info (file modes, etc) which FileInfo doesn't give us
+        struct stat src_stat;
+        if (lstat(src_file, &src_stat) != 0)
+        {
+            result.m_ReturnCode = -1;
+            snprintf(tmpBuffer, sizeof(tmpBuffer), "The properties of source file %s could not be retrieved: %s", src_file, strerror(errno));
+            break;
+        }
+
+        if (S_ISDIR(src_stat.st_mode))
+        {
+            result.m_ReturnCode = -1;
+            snprintf(tmpBuffer, sizeof(tmpBuffer), "The source path %s is a directory, which is not supported.", src_file);
+            break;
+        }
+
+        FileInfo dst_file_info = StatCacheStat(stat_cache, target_file);
+        if (dst_file_info.Exists())
+        {
+            if (dst_file_info.IsDirectory())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists as a directory.", target_file);
+                break;
+            }
+
+            if (dst_file_info.IsReadOnly())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists and is read-only.", target_file);
+                break;
+            }
+        }
+
+        if ((src_stat.st_mode & S_IFMT) == S_IFLNK)
+        {
+            // It's a symlink
+
+            // Add 2 bytes, 1 for the null character and one so we can tell that the readlink return value did not get truncated
+            size_t bufferSize = src_stat.st_size + 2; 
+            char* link_contents = static_cast<char*>(LinearAllocate(&scratch, bufferSize, 1));
```

Comment:
I think `link_target` (or maybe just `target`) would be a better name for this.

Reply:
Done.